### PR TITLE
fnm: use git strategy for livecheck

### PR DESCRIPTION
**Pull request by vladimyr**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

Using `GithubLatest` is not necessary here.

### CODE REVIEW

File: Formula/fnm.rb [hunk 1]
```diff
@@ -6,11 +6,6 @@ class Fnm < Formula
   license "GPL-3.0-only"
   head "https://github.com/Schniz/fnm.git"
 
-  livecheck do
-    url :stable
-    strategy :github_latest
-  end
-
```
Comment (samford) [inline Formula/fnm.rb:13]:
```suggestion
  livecheck do
    url :stable
    regex(/^v?(\d+(?:\.\d+)+)$/i)
  end

```

**Review by samford — CHANGES_REQUESTED** [review]
I agree that `GithubLatest` isn't needed for `fnm`. Looking at the tags, they reliably use a format like `v1.2.3` and there's one `v1.22.0-beta-1` tag (which gets omitted for containing `beta`, one of the `UNSTABLE_VERSION_KEYWORDS`).

However, it wouldn't hurt to use the standard regex for Git tags like `1.2.3`/`v1.2.3`, as this would omit unexpected tags in the future that can't be accounted for using the `UNSTABLE_VERSION_KEYWORDS`. Since we know the tag format, it's easy to do this now instead of waiting for the default check to potentially break over time.

**Comment by vladimyr** [comment]
Isn't the optional `v` prefix already covered by Git strategy?
https://github.com/Homebrew/brew/blob/3.2.0/Library/Homebrew/livecheck/strategy/git.rb#L115-L130 (talking about line 124 specifically)

> However, it wouldn't hurt to use the standard regex for Git tags like `1.2.3`/`v1.2.3`, as this would omit unexpected tags in the future that can't be accounted for using the `UNSTABLE_VERSION_KEYWORDS`. Since we know the tag format, it's easy to do this now instead of waiting for the default check to potentially break over time.

While I do agree with this I fail to see how it is `fnm` specific because the majority of Github releases use given tag format. It should be the default pattern then instead of copying this thing over and over? 😕

**Comment by samford** [comment]
> Isn't the optional `v` prefix already covered by Git strategy?
> https://github.com/Homebrew/brew/blob/3.2.0/Library/Homebrew/livecheck/strategy/git.rb#L115-L130 (talking about line 124 specifically)

The default regex for the `Git` strategy captures everything after the first digit in the tag so, yes, a leading `v` wouldn't be part of the capture group. I hinted at it and didn't say it outright but the default `Git` behavior works fine for `fnm` at the moment and my suggestion of adding a regex is intended to formalize the tag format and avoid potential issues in the future.

By adding the standard regex for Git tags like `1.2.3`/`v1.2.3`, it will restrict matching to only the tags that we're interested in. The default `Git` behavior without a regex can lead to issues around leading text (e.g., matching an unrelated version due to a [non-numeric] prefix being omitted) and trailing text (e.g., tags including extra text that won't/can't be accounted for by `UNSTABLE_VERSION_KEYWORDS`).

---

> While I do agree with this I fail to see how it is `fnm` specific because the majority of Github releases use given tag format. It should be the default pattern then instead of copying this thing over and over? 😕

Sorry in advance for the lengthy explanation but there's quite a bit to explain about this topic. Suffice it to say, I thought about doing this in the past but I decided against it after weighing the evidence.

The reason why we don't use the standard regex for Git tags like `1.2.3`/`v1.2.3` as the default regex in the `Git` strategy is that it only matches a very specific tag format and quietly omits everything else. Yes, many repositories use this format but I've seen a lot of unexpected variation despite that.

With a restrictive default, we would miss tags that we should be matching and, in most cases, we wouldn't even know this is happening at a glance. Omissions due to a restrictive default regex would only become noticeable when the formula is updated to a version that livecheck doesn't match (i.e., the current version becomes red).

From the standpoint of livecheck, it's preferable for the `Git` strategy to have a permissive/loose default as it will fail by matching _more_ than it should (not less) and this is more readily apparent in livecheck's output. These issues can be identified at a glance, so we can more easily see which formulae need work.

The only benefit of using the standard `Git` regex as the default is that we could omit it for the formulae where it applies correctly. In my view, the detriments heavily outweigh this and there's value in verifying that a `Git` regex is appropriate before applying it. That's what we do now and a regex in a livecheck block using the `Git` strategy generally means, "I checked the repository's tags and verified that this regex was correct at the time it was added".

---

I would actually argue that the `Git` strategy's regex should be made even looser (e.g., only omitting an optional leading `v` by default). My reasoning is that this would also make it readily apparent when a repository contains tags that use a prefix and force us to account for it in a regex.

Some repositories contain multiple pieces of software and the current default `Git` behavior causes problems because it's prefix agnostic. That is to say, omitting any leading non-digit text means that prefixes are ignored and versions are mixed together. I've seen livecheck reporting a version from unrelated software a number of times before and it's not always easy to spot. It's better for this to fail in a way that we can easily see and fix, rather than doing something that looks plausible but is incorrect.

However, before I loosen the `Git` strategy's default regex, I want to add `livecheck` blocks to any formulae that would be affected. I haven't researched how much effort this would require but it's already on my to-do list (after higher-priority items).

---

In my view, the ideal `Git` situation would be to have an appropriate regex in each formula using the strategy (or something equivalent), rather than assuming the default behavior is correct. There are currently 2500+ formulae using `Git` (900+ casks), so that's unlikely to happen anytime soon, if ever. Instead, I chip away at this by adding a regex to `Git` formulae as I check the tags to determine what regex is appropriate, as in this PR.

That said, I should probably be clear that what I've said above is specific to the `Git` strategy. Some of the other strategies are very domain-specific and the defaults can be appropriate for almost all related formulae without requiring a `livecheck` block. `Git` is different because the variation in tags across thousands of repositories is something that we can only reliably account for in individual `livecheck` blocks after manual verification.

---

Past that, it may be worth mentioning that the ergonomics of this will improve a bit sometime in the not-too-distant future. Right now it's necessary to copy and paste common regexes, which isn't a great situation. I personally minimize this pain by using snippets (i.e., I type a keyword and it gets automatically replaced) but it's a hurdle for anyone else.

I have some work stashed locally that creates constants for common regexes (like this one) and regex parts. The goal is to have something simple that people can remember and type, as well as to make it clear what's typical. I need to revisit this and wrap it up into a PR but there are still some other items on my to-do list that I need to work through first.

**Review by samford — APPROVED** [review]
Thanks, @vladimyr.

**Pull request merged**